### PR TITLE
graceful-grasshoppers

### PR DESCRIPTION
**Pull request by georgemunyoro**

### CODE REVIEW

**Comment by SebastiaanZ** [comment]
This Pull Request contains commits with a timestamp after the deadline. We'll need to have a look at that before merging this PR.


flake8 lint results
```
 ./graceful-grasshoppers/PySpace/PySpace/settings.py:100:110: E231 missing whitespace after ','
./graceful-grasshoppers/PySpace/PySpace/settings.py:148:12: E222 multiple spaces after operator
./graceful-grasshoppers/PySpace/api/urls.py:1:1: F401 'django.contrib.admin' imported but unused
./graceful-grasshoppers/PySpace/api/tests.py:1:1: F401 'django.test.TestCase' imported but unused
./graceful-grasshoppers/PySpace/api/serializers.py:4:1: E302 expected 2 blank lines, found 1
./graceful-grasshoppers/PySpace/posts/urls.py:5:1: W191 indentation contains tabs
./graceful-grasshoppers/PySpace/posts/urls.py:6:1: W191 indentation contains tabs
./graceful-grasshoppers/PySpace/posts/urls.py:7:1: W191 indentation contains tabs
./graceful-grasshoppers/PySpace/posts/urls.py:8:1: W191 indentation contains tabs
./graceful-grasshoppers/PySpace/posts/urls.py:9:1: W191 indentation contains tabs
./graceful-grasshoppers/PySpace/posts/urls.py:10:1: W191 indentation contains tabs
./graceful-grasshoppers/PySpace/posts/urls.py:11:1: W191 indentation contains tabs
./graceful-grasshoppers/PySpace/posts/views.py:73:16: F821 undefined name 'Response'
./graceful-grasshoppers/PySpace/posts/views.py:198:1: W391 blank line at end of file
./graceful-grasshoppers/PySpace/posts/tests.py:1:1: F401 'django.test.TestCase' imported but unused
./graceful-grasshoppers/PySpace/users/urls.py:1:1: F401 'django.urls.include' imported but unused
./graceful-grasshoppers/PySpace/users/admin.py:2:1: F401 'django.contrib.auth.get_user_model' imported but unused
./graceful-grasshoppers/PySpace/users/views.py:111:5: F841 local variable 'payload' is assigned to but never used
./graceful-grasshoppers/PySpace/users/tests.py:1:1: F401 'django.test.TestCase' imported but unused
./graceful-grasshoppers/PySpace/users/serializers.py:6:1: F401 'api.serializers as api_serializers' imported but unused
./shiny_sheep/shiny_sheep/chat/models.py:1:1: F401 'django.db.models' imported but unused
./shiny_sheep/shiny_sheep/chat/admin.py:1:1: F401 'django.contrib.admin' imported but unused
./shiny_sheep/shiny_sheep/chat/routing.py:10:1: W391 blank line at end of file
./shiny_sheep/shiny_sheep/chat/views.py:1:1: F401 'django.shortcuts.render' imported but unused
./shiny_sheep/shiny_sheep/chat/tests.py:1:1: F401 'django.test.TestCase' imported but unused
./shiny_sheep/shiny_sheep/chat/consumers.py:8:1: E302 expected 2 blank lines, found 1
./shiny_sheep/shiny_sheep/chat/consumers.py:51:1: E302 expected 2 blank lines, found 1
```

**Comment by SebastiaanZ** [comment]
Notes for reviewer:
- This team had a couple of late merge commits (up to 17 minutes after the Jam). The actual commits merged by these commits were committed within the timeframe of the jam.

- This team used a GitHub Actions file to auto-format the code with Black. It was removed to prevent it from being triggered on other submissions, but this was the contents:

```yaml
on: [pull_request]
jobs:
  format:
    runs-on: ubuntu-latest
    steps:
      - name: Black Code Formatter
        uses: lgeiger/black-action@v1.0.1
```

File: graceful-grasshoppers/README.md [hunk 1]
```diff
@@ -1,2 +1,114 @@
...
+```
+Now you have your virtual environment activated. (to be sure, make sure the prompt is now `(pyspace) the/path/of/the/current/folder`)
+4. Install the requirements.
+```
+pip install -r requirements.txt
+```
+5. Make Migrations.
+- MacOS or Linux (bash): 
+```
+python3 makemigrations
```
Comment (kosayoda) [inline graceful-grasshoppers/README.md:36]: This should be
```
python3 manage.py makemigrations
```

File: graceful-grasshoppers/README.md [hunk 2]
```diff
@@ -1,2 +1,114 @@
...
+```
+- Windows (cmd): 
+```
+py -m venv pyspace
+.\venv\Scripts\activate
+```
+Now you have your virtual environment activated. (to be sure, make sure the prompt is now `(pyspace) the/path/of/the/current/folder`)
+4. Install the requirements.
+```
+pip install -r requirements.txt
```
Comment (kosayoda) [inline graceful-grasshoppers/README.md:31]: Alternatively, consider using a dependency management tool such as `pipenv`, which would also allow for custom run scripts, which would simplify some installation steps.

**Review by kosayoda — COMMENTED** [review]
Thank you for participating in this code jam!

I was unable to get the site running. Trying to make the migrations results in a `django.db.utils.OperationalError: no such table: api_file`.

Code quality is decent. The docstrings could use some addtional information, such as return values. Additionally, consider type hints to add clarification.

The project uses Vue for the frontend. I did not review the Vue/JS code as it is out of scope and I have no experience with those languages. However, one thing I was confused about was the use of 2 space indents for JS but tabs for Vue. Consistency might ease development here.

Another odd thing was that the `requirements.txt` file encoding was UTF-16 little endian + CRLF line terminators, while the majority of the project is UTF-8.

However, from the screenshots provided, the website looks very promising and thematic. It is unfortunate that the website could not be ran. I wish you all the best in the future and hope you join other code jams!

**Pull request merged**